File: frontend/src/services/recommendations.ts
/**
 * Recommendations Service
 * Service for getting career recommendations with various endpoints
 */
import apiClient from './apiClient';
import { BaseResponse } from './types';

/**
 * Recommendation request types
 */
export interface RecommendationConstraints {
  min_wage?: number;
  remote_preferred?: boolean;
  max_education_level?: number;
  [key: string]: any;
}

export interface RecommendationRequest {
  skills?: string[];
  skill_importance?: Record<string, number>; // 0-5 scale
  interests?: Record<string, number>; // RIASEC categories, 0-7 scale
  work_values?: Record<string, number>; // 0-7 scale
  constraints?: RecommendationConstraints;
  top_n?: number; // 1-20, default 5
  use_ml?: boolean; // default true
}

/**
 * Recommendation explanation types
 */
export interface TopFeature {
  feature: string;
  contribution: number;
  user_value: number;
  career_value: number;
}

export interface RecommendationExplanation {
  method: string;
  confidence: 'Low' | 'Med' | 'High';
  top_contributing_skills?: TopFeature[];
  why_points?: string[];
  [key: string]: any;
}

export interface OpenAIEnhancement {
  why_narrative?: string;
  refined_explanation?: string;
  [key: string]: any;
}

/**
 * Why narrative structure (can be string or object)
 */
export interface WhyNarrative {
  primary?: string;
  points?: string[];
  summary?: string;
  enhanced?: string;
  top_features?: string[];
  next_steps?: string;
  [key: string]: any;
}

/**
 * Single recommendation response
 */
export interface CareerRecommendation {
  career_id: string;
  name: string;
  soc_code: string;
  score: number;
  confidence: 'Low' | 'Med' | 'High';
  explanation: RecommendationExplanation;
  outlook?: Record<string, any>;
  education?: Record<string, any>;
  openai_enhancement?: OpenAIEnhancement;
  // Enhanced format fields
  confidence_band?: {
    level: 'Low' | 'Med' | 'High';
    score_range: [number, number];
  };
  explainability?: {
    top_features: TopFeature[];
    summary: string;
  };
  why?: string | WhyNarrative;
  [key: string]: any;
}

/**
 * Legacy recommendations response
 */
export interface RecommendationsResponse {
  recommendations: CareerRecommendation[];
  user_features: {
    num_skills_provided: number;
    interests_provided: boolean;
    values_provided: boolean;
    constraints_provided: boolean;
  };
  method: 'ml_model' | 'baseline';
}

/**
 * Enhanced recommendations response
 */
export interface EnhancedRecommendationsResponse {
  careers: CareerRecommendation[];
  alternatives: CareerRecommendation[];
  method: 'ml_model' | 'baseline';
  user_features: {
    num_skills_provided: number;
    interests_provided: boolean;
    values_provided: boolean;
    constraints_provided: boolean;
  };
}

/**
 * Model status response
 */
export interface ModelStatusResponse {
  model_loaded: boolean;
  scaler_loaded: boolean;
  model_version: string | null;
  fallback_to_baseline: boolean;
}

/**
 * Simple recommendations query parameters
 */
export interface SimpleRecommendationsParams {
  skills?: string; // Comma-separated list of skills
  top_n?: number; // 1-20, default 5
}

/**
 * POST /api/recommendations/recommendations - Enhanced recommendations
 * Returns 3-5 careers with confidence bands, explainability, alternatives, and why narrative
 */
export const getEnhancedRecommendations = async (
  request: RecommendationRequest
): Promise<EnhancedRecommendationsResponse> => {
  const response = await apiClient.post<BaseResponse<EnhancedRecommendationsResponse>>(
    '/api/recommendations/recommendations',
    {
      skills: request.skills,
      skill_importance: request.skill_importance,
      interests: request.interests,
      work_values: request.work_values,
      constraints: request.constraints,
      top_n: request.top_n ?? 5,
      use_ml: request.use_ml ?? true,
    }
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get enhanced recommendations');
  }

  return response.data.data!;
};

/**
 * POST /api/recommendations/recommend - Legacy recommendations
 * Returns career recommendations based on user skills, interests, values, and constraints
 */
export const getRecommendations = async (
  request: RecommendationRequest
): Promise<RecommendationsResponse> => {
  const response = await apiClient.post<BaseResponse<RecommendationsResponse>>(
    '/api/recommendations/recommend',
    {
      skills: request.skills,
      skill_importance: request.skill_importance,
      interests: request.interests,
      work_values: request.work_values,
      constraints: request.constraints,
      top_n: request.top_n ?? 5,
      use_ml: request.use_ml ?? true,
    }
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get recommendations');
  }

  return response.data.data!;
};

/**
 * GET /api/recommendations/recommend/simple - Simple recommendations (query params)
 * Quick endpoint for testing - just provide skills as comma-separated string
 */
export const getSimpleRecommendations = async (
  params: SimpleRecommendationsParams
): Promise<RecommendationsResponse> => {
  const queryParams: Record<string, string | number> = {};

  if (params.skills) {
    queryParams.skills = params.skills;
  }

  if (params.top_n !== undefined) {
    queryParams.top_n = params.top_n;
  }

  const response = await apiClient.get<BaseResponse<RecommendationsResponse>>(
    '/api/recommendations/recommend/simple',
    {
      params: queryParams,
    }
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get simple recommendations');
  }

  return response.data.data!;
};

/**
 * GET /api/recommendations/model/status - Model status check
 * Returns status of the ML model - whether it's loaded and what version
 */
export const getModelStatus = async (): Promise<ModelStatusResponse> => {
  const response = await apiClient.get<BaseResponse<ModelStatusResponse>>(
    '/api/recommendations/model/status'
  );

  if (!response.data.success) {
    throw new Error(response.data.message || 'Failed to get model status');
  }

  return response.data.data!;
};

/**
 * Recommendations service object with all methods
 */
export const recommendationsService = {
  getEnhancedRecommendations,
  getRecommendations,
  getSimpleRecommendations,
  getModelStatus,
};

export default recommendationsService;

File: backend/routes/recommendations.py
"""
API routes for career recommendations
"""
from fastapi import APIRouter, HTTPException, status, Query, Body
from models.schemas import BaseResponse, ErrorResponse
from services.recommendation_service import CareerRecommendationService
from utils.security import sanitize_error_message
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field

router = APIRouter()
recommendation_service = CareerRecommendationService()

# Try to load model on startup
recommendation_service.load_model_artifacts()


class RecommendationRequest(BaseModel):
    """Request schema for career recommendations"""
    skills: Optional[List[str]] = Field(None, description="List of skill names")
    skill_importance: Optional[Dict[str, float]] = Field(
        None, 
        description="Dict mapping skill names to importance scores (0-5)"
    )
    interests: Optional[Dict[str, float]] = Field(
        None,
        description="Dict mapping RIASEC categories to scores (0-7). Categories: Realistic, Investigative, Artistic, Social, Enterprising, Conventional"
    )
    work_values: Optional[Dict[str, float]] = Field(
        None,
        description="Dict mapping work values to scores (0-7). Values: Achievement, Working Conditions, Recognition, Relationships, Support, Independence"
    )
    constraints: Optional[Dict[str, Any]] = Field(
        None,
        description="Dict with constraints like min_wage, remote_preferred, max_education_level"
    )
    top_n: int = Field(5, ge=1, le=20, description="Number of recommendations to return")
    use_ml: bool = Field(True, description="Whether to use ML model or baseline ranking")


@router.post("/recommendations", response_model=BaseResponse)
async def get_recommendations(request: RecommendationRequest = Body(...)):
    """
    Get enhanced career recommendations with:
    - 3-5 careers
    - Confidence bands (with score ranges)
    - Explainability (top features)
    - Alternatives
    - "Why" narrative
    
    Uses ML model if available, otherwise falls back to baseline similarity ranking
    """
    try:
        result = recommendation_service.get_enhanced_recommendations(
            skills=request.skills,
            skill_importance=request.skill_importance,
            interests=request.interests,
            work_values=request.work_values,
            constraints=request.constraints,
            use_ml=request.use_ml,
            use_openai=True
        )
        
        return BaseResponse(
            success=True,
            message=f"Generated {len(result['careers'])} recommendations with {len(result.get('alternatives', []))} alternatives",
            data=result
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to generate recommendations",
                error=sanitize_error_message(e)
            ).model_dump()
        )


@router.post("/recommend", response_model=BaseResponse)
async def get_recommendations_legacy(request: RecommendationRequest = Body(...)):
    """
    Legacy endpoint - Get career recommendations based on user skills, interests, values, and constraints
    
    Uses ML model if available, otherwise falls back to baseline similarity ranking
    """
    try:
        result = recommendation_service.recommend(
            skills=request.skills,
            skill_importance=request.skill_importance,
            interests=request.interests,
            work_values=request.work_values,
            constraints=request.constraints,
            top_n=5,
            use_ml=request.use_ml
        )
        
        return BaseResponse(
            success=True,
            message=f"Generated {len(result['recommendations'])} recommendations",
            data=result
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to generate recommendations",
                error=sanitize_error_message(e)
            ).model_dump()
        )


@router.get("/recommend/simple", response_model=BaseResponse)
async def get_simple_recommendations(
    skills: Optional[str] = Query(None, description="Comma-separated list of skills"),
    top_n: int = Query(5, ge=1, le=20, description="Number of recommendations")
):
    """
    Simple recommendation endpoint - just provide skills as comma-separated string
    Useful for quick testing
    """
    try:
        skills_list = [s.strip() for s in skills.split(",")] if skills else None
        
        result = recommendation_service.recommend(
            skills=skills_list,
            top_n=top_n,
            use_ml=True
        )
        
        return BaseResponse(
            success=True,
            message=f"Generated {len(result['recommendations'])} recommendations",
            data=result
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to generate recommendations",
                error=sanitize_error_message(e)
            ).model_dump()
        )


@router.get("/model/status", response_model=BaseResponse)
async def get_model_status():
    """
    Get status of the ML model - whether it's loaded and what version
    """
    try:
        has_model = recommendation_service.ml_model is not None
        has_scaler = recommendation_service.scaler is not None
        
        return BaseResponse(
            success=True,
            message="Model status retrieved",
            data={
                "model_loaded": has_model,
                "scaler_loaded": has_scaler,
                "model_version": recommendation_service.model_version,
                "fallback_to_baseline": not has_model
            }
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=ErrorResponse(
                success=False,
                message="Failed to get model status",
                error=str(e)
            ).model_dump()
        )





